libsolidity/codegen/ArrayUtils.cpp

```diff
@@ -358,27 +351,23 @@ void ArrayUtils::copyArrayToMemory(ArrayType const& _sourceType, bool _padToWord
 			m_context << Instruction::SWAP1 << u256(32) << Instruction::ADD;
 			m_context << Instruction::SWAP1;
 		}
-		// convert length to size
-		if (baseSize > 1)
-			m_context << u256(baseSize) << Instruction::MUL;
+		if (!_sourceType.isByteArray())
+			convertLengthToSize(_sourceType);
 		// stack: <target> <source> <size>
 		m_context << Instruction::DUP1 << Instruction::DUP4 << Instruction::DUP4;
 		// We can resort to copying full 32 bytes only if
 		// - the length is known to be a multiple of 32 or
 		// - we will pad to full 32 bytes later anyway.
-		if (((baseSize % 32) == 0) || _padToWordBoundaries)
+		if (!_sourceType.isByteArray() || _padToWordBoundaries)
 			utils.memoryCopy32();
 		else
 			utils.memoryCopy();
 
 		m_context << Instruction::SWAP1 << Instruction::POP;
 		// stack: <target> <size>
 
-		bool paddingNeeded = false;
-		if (_sourceType.isDynamicallySized())
-			paddingNeeded = _padToWordBoundaries && ((baseSize % 32) != 0);
-		else
-			paddingNeeded = _padToWordBoundaries && (((_sourceType.length() * baseSize) % 32) != 0);
+		bool paddingNeeded = _padToWordBoundaries && _sourceType.isByteArray();
```

Comment:
This covers both the other cases it replaces? Could you explain why?

Reply:
Those are @chriseth's changes actually - I haven't second-guessed them yet, nor spent too much time thinking about them...